pkg/proxy/ipvs/graceful_delete.go

```diff
@@ -0,0 +1,250 @@
+}
+
+// GracefulDeleteRSQueueQueue is a priority heap where the lowest ProcessAt is at the front of the queue
+type GracefulDeleteRSQueueQueue []*GracefulDeleteRS
+
+// Len is the length of the queue.
+func (h GracefulDeleteRSQueueQueue) Len() int { return len(h) }
+
+// Less returns true if queue[i] < queue[j].
+func (h GracefulDeleteRSQueueQueue) Less(i, j int) bool { return h[i].ProcessAt.Before(h[j].ProcessAt) }
+
+// Swap swaps index i and j.
+func (h GracefulDeleteRSQueueQueue) Swap(i, j int) { h[i], h[j] = h[j], h[i] }
+
+// Push a new TimedValue on to the queue.
+func (h *GracefulDeleteRSQueueQueue) Push(x interface{}) {
+	*h = append(*h, x.(*GracefulDeleteRS))
+}
+
+// Pop the lowest ProcessAt item.
+func (h *GracefulDeleteRSQueueQueue) Pop() interface{} {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1]
+	return x
+}
+
+type UniqueQueue struct {
+	lock  sync.Mutex
+	queue GracefulDeleteRSQueueQueue
+	set   sets.String
+}
+
+// Add push an new element to the queue
+func (q *UniqueQueue) Add(rs GracefulDeleteRS) bool {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	uniqueRS := rs.String()
+	if q.set.Has(uniqueRS) {
+		return false
+	}
+	glog.V(5).Infof("Pushing rs %v to graceful delete queue: %+v", rs)
+	heap.Push(&q.queue, &rs)
+	q.set.Insert(uniqueRS)
+	return true
+}
+
+// Remove remove an element from the queue
+func (q *UniqueQueue) Remove(rs *GracefulDeleteRS) bool {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	uniqueRS := rs.String()
+	if !q.set.Has(uniqueRS) {
+		return false
+	}
+	q.set.Delete(uniqueRS)
+	for i, val := range q.queue {
+		if val.String() == uniqueRS {
+			heap.Remove(&q.queue, i)
+			return true
+		}
+	}
+	return true
+}
+
+// Head return the first element from the queue
+func (q *UniqueQueue) Head() (*GracefulDeleteRS, bool) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	if len(q.queue) == 0 {
+		return nil, false
+	}
+	result := q.queue[0]
+	return result, true
+}
+
+func (q *UniqueQueue) Pop() (*GracefulDeleteRS, bool) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	if len(q.queue) == 0 {
+		return nil, false
+	}
+	result := heap.Pop(&q.queue).(*GracefulDeleteRS)
+	q.set.Delete(result.String())
+	return result, true
+}
+
+// Exist check whether the specified unique RS is in the queue
+func (q *UniqueQueue) Exist(uniqueRS string) (*GracefulDeleteRS, bool) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	if !q.set.Has(uniqueRS) {
+		return nil, false
+	}
+	for _, val := range q.queue {
+		if val.String() == uniqueRS {
+			return val, true
+		}
+	}
+	return nil, false
+}
+
+type GracefulDeleteManager struct {
+	queue UniqueQueue
+	ipvs  utilipvs.Interface
+}
+
+func NewGracefulDeleteManager(ipvs utilipvs.Interface) *GracefulDeleteManager {
+	return &GracefulDeleteManager{
+		queue: UniqueQueue{
+			queue: GracefulDeleteRSQueueQueue{},
+			set:   sets.NewString(),
+		},
+		ipvs: ipvs,
+	}
+}
+
+func (m *GracefulDeleteManager) RSInGracefulDelete(uniqueRS string) bool {
```

Comment:
Can we call this PendingGracefulTermination?

Reply:
This function to check whether rs is in graceful termination list. I think `InTerminationList` would be better.